Cleanup Rust interface for Target::get_non_global_operation_names()

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit cleans up the rust interface for the
`Target::get_non_global_operation_names()` method. The function was requiring a mutable reference to the target for querying the data from rust which is a mismatch from expectations. This was because the function was trying to cache the result which was a Python-ism that was faithfully ported when the Target moved from Python to Rust, but this doesn't really fit in a rust interface because a rust user takes more care around mutability. The caching is a construct for Python and doesn't really fit with a rust interface. This commit updates the rust interface to not cache this data and move it to Python. It's arguable whether a cache is actually needed for this or not, but just out of an abundance of caution this leaves it in place. However, the cache is moved out of the Rust struct to avoid carrying the dead weight around in a standalone C/Rust context.

### Details and comments

mtreinish committed:
Cleanup Rust interface for Target::get_non_global_operation_names()

This commit cleans up the rust interface for the
Target::get_non_global_operation_names() method. The function was
requiring a mutable reference to the target for querying the data from
rust which is a mismatch from expectations. This was because the
function was trying to cache the result which was a Python-ism that was
faithfully ported when the Target moved from Python to Rust, but this
doesn't really fit in a rust interface because a rust user takes more
care around mutability. The caching is a construct for Python and
doesn't really fit with a rust interface. This commit updates the rust
interface to not cache this data and move it to Python. It's arguable
whether a cache is actually needed for this or not, but just out of an
abundance of caution this leaves it in place. However, the cache is
moved out of the Rust struct to avoid carrying the deadweight around in
a standalone C/Rust context.

Cryoris commented on qiskit/transpiler/target.py:
```diff
@@ -276,6 +278,37 @@ def __init__(
+                    strict_direction
+                )
+            return self._non_global_basis_strict
+        if getattr(self, "_non_global_basis", None) is None:
```
Is there a reason to use `getattr` over just this? 
```suggestion
            if self._non_global_basis_strict is None:
                self._non_global_basis_strict = super()._get_non_global_operation_names(
                    strict_direction
                )
            return self._non_global_basis_strict
        if self._non_global_basis is None:
```

mtreinish replied:
There were code paths that were apparently skipping `__init__` that called this method (given how subclassing works with PyO3 I could see that being the case). When I wrote this locally I had it as you suggested and was getting `AttributeErrors` in a few tests because the attribute hadn't been populated yet. Doing the getattr covered that edge case.

Cryoris replied:
It might be nice to leave a comment explaining this for our future selves 😄

jakelishman replied:
or possibly fix the places that are calling methods that require initialisation without actually initialising the objects... That's a huge footgun waiting to go off further.

raynelfss replied:
@mtreinish which tests are failing? Is it possible that it might have something to do with `FakeTarget`? I don't think so since it calls `__init__`, but it doesn't call `__new__`.

mtreinish replied:
It were some transpile tests that were failing in the basis translator. I think it was related to the `_FakeTarget`. I fixed the issue in: https://github.com/Qiskit/qiskit/pull/14921/commits/19b890bf67ef7f4979958028537c0175742c69ec and used the more normal syntax without the getattr in that commit.

mtreinish committed:
Fix Target python object construction

This commit fixes the object creation path of the Target so there isn't
a path for the target to be created without it's Python space caches
getting created.

mtreinish commented on qiskit/transpiler/target.py:
```diff
@@ -258,24 +258,43 @@ def __new__(  # pylint: disable=keyword-arg-before-vararg
-        acquire_alignment=1,  # pylint: disable=unused-argument
-        qubit_properties=None,  # pylint: disable=unused-argument
-        concurrent_measurements=None,  # pylint: disable=unused-argument
-    ):
```
I need to restore this to fix subclasses, I forgot that the Target needs to be subclass safe.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14921/commits/f86cbe44013a97c7eea0c9e4aff5e4932be2e7cf

mtreinish committed:
Merge remote-tracking branch 'origin/main' into no-mutable-target

mtreinish committed:
Fix subclassing

jakelishman commented on crates/transpiler/src/target/mod.rs:
```diff
@@ -1214,30 +1188,17 @@ impl Target {
-        Some(self.generate_non_global_op_names(strict_direction))
+    pub fn get_non_global_operation_names(&mut self, strict_direction: bool) -> Vec<&str> {
+        self.generate_non_global_op_names(strict_direction)
     }
```
This no longer needs to take `&mut self`, but secondly, now there's no caching, can we delete get rid of one of these two duplicate methods?  E.g. just have `get_non_global_operation_names`.

mtreinish replied:
I made it `&self` and removed one of the functions in: https://github.com/Qiskit/qiskit/pull/14921/commits/fe6253e3c7a3ca0fd9acbf765b2889a26b4b5645

jakelishman commented on qiskit/transpiler/target.py:
```diff
@@ -271,11 +279,36 @@ def __init__(
-        self._instruction_schedule_map = None
+        super().__init__()
+
+    def get_non_global_operation_names(self, strict_direction=False):
+        """Return the non-global operation names for the target
+
+        The non-global operations are those in the target which don't apply
+        on all qubits (for single qubit operations) or all multi-qubit qargs
+        (for multi-qubit operations).
+
+        Args:
+            strict_direction (bool): If set to ``True`` the multi-qubit
+                operations considered as non-global respect the strict
+                direction (or order of qubits in the qargs is significant). For
+                example, if ``cx`` is defined on ``(0, 1)`` and ``ecr`` is
+                defined over ``(1, 0)`` by default neither would be considered
+                non-global, but if ``strict_direction`` is set ``True`` both
+                ``cx`` and ``ecr`` would be returned.
+
+        Returns:
+            List[str]: A list of operation names for operations that aren't global in this target
+        """
+        if strict_direction:
+            if self._non_global_basis_strict is None:
+                self._non_global_basis_strict = super()._get_non_global_operation_names(
+                    strict_direction
+                )
+            return self._non_global_basis_strict
+        if self._non_global_basis is None:
+            self._non_global_basis = super()._get_non_global_operation_names(strict_direction)
+        return self._non_global_basis
```
Do we need to cache this?  I'm concerned that we no longer have any cache invalidation (not sure if we did before tbh).

mtreinish replied:
I wasn't sure, so I left it in out an abundance of caution. I did mirror the invalidation that was there before, we clear the cache when we add an instruction. There wasn't anything else that I could find.

jakelishman replied:
Oh sorry, I misread the cache invalidation below as being in the initialiser of `FakeTarget`.  In that case I'm fine with it.

jakelishman commented on crates/transpiler/src/target/mod.rs:
```diff
@@ -699,16 +695,9 @@ impl Target {
-            .into_pyobject(py)?
-            .unbind())
+    #[pyo3(name = "_get_non_global_operation_names", signature = (/, strict_direction=false,))]
+    fn py_get_non_global_operation_names(&mut self, strict_direction: bool) -> Vec<&str> {
```
This also doesn't need to take `&mut self`, though I suppose it's not quite as important for the Python-space wrapper.

mtreinish replied:
I switched it anyway in: https://github.com/Qiskit/qiskit/pull/14921/commits/fe6253e3c7a3ca0fd9acbf765b2889a26b4b5645

mtreinish committed:
Merge remote-tracking branch 'origin/main' into no-mutable-target

mtreinish committed:
Simplify rust interface
